<!-- 
  Turtlebot navigation simulation:
  - stage
  - map_server
  - move_base
  - static map
  - amcl
  - rviz view
 -->
<launch>
  <arg name="base"       default="$(optenv TURTLEBOT_BASE create)"/>  <!-- create, rhoomba -->
  <arg name="stacks"     default="$(optenv TURTLEBOT_STACKS circles)"/>  <!-- circles, hexagons -->
  <arg name="3d_sensor"  default="$(optenv TURTLEBOT_3D_SENSOR kinect)"/>  <!-- kinect, asus_xtion_pro -->

  <!-- Name of the map to use (without path nor extension) and initial position -->
  <arg name="map_file"       default=" $(find collvoid_turtlebot)/world/my_map.yaml"/> <!-- robopark_plan -->
  <arg name="world_file"     default=" $(find collvoid_turtlebot)/world/my_map.world"/>
  <arg name="initial_pose_x" default="2.1"/>
  <arg name="initial_pose_y" default="2.1"/>
  <arg name="initial_pose_a" default="3.14"/>

  <param name="/use_sim_time" value="true"/>
  <!--  ******************** Stage ********************  -->
  <!-- 
        Publishes transforms:
          /base_link -> /base_laser
          /base_footprint -> /base_link (identity)
          /odom -> base_footprint
        Publishes topics:
          /odom : odometry data from the simulated odometry
          /base_scan : laser data from the simulated laser
          /base_pose_ground_truth : the ground truth pose
        Parameters:
          base_watchdog_timeout : time (s) after receiving the last command on cmd_vel before stopping the robot
        Args:
          -g : run in headless mode.
  --> 
  <node pkg="stage_ros" type="stageros" name="stageros" args="$(arg world_file)">
    <param name="base_watchdog_timeout" value="0.5"/>
    <remap from="odom" to="odom"/>
    <remap from="base_pose_ground_truth" to="base_pose_ground_truth"/>
    <remap from="cmd_vel" to="mobile_base/commands/velocity"/>
    <remap from="base_scan" to="scan"/>
  </node>

  <!--  ***************** Robot Model *****************  -->
  <include file="$(find turtlebot_bringup)/launch/includes/robot.launch.xml">
    <arg name="base" value="$(arg base)" />
    <arg name="stacks" value="$(arg stacks)" />
    <arg name="3d_sensor" value="$(arg 3d_sensor)" />
  </include>
  <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher">
    <param name="use_gui" value="true"/>
  </node>

  <!-- Command Velocity multiplexer -->
  <node pkg="nodelet" type="nodelet" name="mobile_base_nodelet_manager" args="manager"/>
  <node pkg="nodelet" type="nodelet" name="cmd_vel_mux" args="load yocs_cmd_vel_mux/CmdVelMuxNodelet mobile_base_nodelet_manager">
    <param name="yaml_cfg_file" value="$(find turtlebot_bringup)/param/mux.yaml"/>
    <remap from="cmd_vel_mux/output" to="mobile_base/commands/velocity"/>
  </node>



  <!--  ****** Maps *****  -->
  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)">
    <param name="frame_id" value="/map"/>
  </node>

 <!--- Run AMCL -->
  <include file="$(find collvoid_turtlebot)/launch/amcl_turtlebot.launch"> 
    <arg name="scan_topic" value="scan"/>
    <arg name="use_map_topic" value="true"/>
    <arg name="initial_pose_x" value="$(arg initial_pose_x)"/>
    <arg name="initial_pose_y" value="$(arg initial_pose_y)"/>
    <arg name="initial_pose_a" value="$(arg initial_pose_a)"/>
  </include>
<!--- Run Controller -->
  <rosparam command="load" file="$(find collvoid_controller)/goals.yaml"/>
  <node pkg="collvoid_controller" type="controllerRobots.py" name="controller_robots" output="screen" />
  <node pkg="collvoid_controller" type="controller.py" name="controller" output="screen" />

<node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <rosparam command="load" file="$(find collvoid_turtlebot)/params/params_turtle.yaml"/>
<rosparam command="load" file="$(find collvoid_turtlebot)/params/collvoid_config.yaml"/>
    <param name="base_local_planner" value="collvoid_local_planner/CollvoidLocalPlanner" />
<param name="base_global_planner" value="collvoid_simple_global_planner/CollvoidSimpleGlobalPlanner"/>
    <remap from="/position_share_in" to="/position_share"/>
    <remap from="/position_share_out" to="/position_share"/> 
    <remap from="/cmd_vel" to="/mobile_base/commands/velocity"/>
  </node>



  <!--  **************** Visualisation ****************  -->
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find turtlebot_stage)/rviz/robot_navigation.rviz"/>
</launch>
